File: leetcode_0003/cpp/leetcode_0003.cpp
//
// Created by wangzy on 2022/3/21.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        //unordered_map<char,int> window;
        //字符表
        vector<int> window(128,0);
        int left = 0,right = 0;
        while(right < s.size()){
            while(right < s.size() && !window[s[right]]){
                window[s[right]] = 1;
                right++;
                if(res < right - left)
                    res = right - left;
            }
            if(right < s.size() && window[s[right]]){
                window[s[left]]--;
                left++;
            }
        }
        return res;
    }
};

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
    string s = "abcabcbb";
    //result
    //3
//    string s = "bbbbbb";
    //result
    //1
//    string s = "pwwkew";
    //result
    //3
    int res = Solution().lengthOfLongestSubstring(s);
    cout<<res<<endl;
}

File: leetcode_0003/java/leetcode_0003.java
public class leetcode_0003 {
    class Solution {
        public int lengthOfLongestSubstring(String s) {
            int res = 0;
            //字符表
            int[] window = new int[128];
            int left = 0,right = 0;
            while(right < s.length()){
                while(right < s.length() && window[s.charAt(right)] == 0){
                    window[s.charAt(right)] = 1;
                    right++;
                    if(right - left > res)
                        res = right - left;
                }
                if(right < s.length() && window[s.charAt(right)] == 1){
                    window[s.charAt(left)] = 0;
                    left++;
                }
            }
            return res;
        }
    }
    public static void main(String[] args) {
        Solution solution = new leetcode_0003().new Solution();
        String s = "abcabcbb";
        //result
        //3
//        String s = "bbbbb";
        //result
        //1
//        String s = "pwwkew";
        //result
        //3
        System.out.println(solution.lengthOfLongestSubstring(s));
    }
}
